### ISSUE TITLE
Large Geant files

### ISSUE THREAD

**Issue by henryiii** [scikit-hep/uproot#201]
When trying to read a large Geant4 file (almost 4 Gb), uproot fails to access the items in the file (it does get the names, though). There are some small changes in the file compared to ones it can read (part of why it is larger), but nothing major. Uproot is trying to access `branch.name`, but branch is `None`. I'll paste the actual error here next time I connect to the cluster I was on.

**Comment by jpivarski** [comment]
It's already a hint, being told that the branch is `None`. That makes it sound like something wrong with "get obj ref" (I don't remember the exact name of the function that deserializes arbitrary objects, such as TTree and TBranch, as opposed to data inside TTrees). If the branch is actually `None`, something that I would think should never happen, the error message is less likely to be useful than an example file. Even if you can't share it publicly or it's too large to be a unit test, it would help for me to download it and look at it, to see how the branch ends up being `None`.

**Comment by henryiii** [comment]
I've obtained permission to send it to you. PM me with instructions on where you want it put. It's sitting on TACC resources at the moment.

**Comment by henryiii** [comment]
Here's the error:

```
---------------------------------------------------------------------------
AttributeError                            Traceback (most recent call last)
<ipython-input-3-5eb72cbc567f> in <module>
      1 d = uproot.open('/home1/02064/henryiii/work_dir/files_mini_real_orig/Det6_mini_real_orig_t0.root')
      2 #d = uproot.open('/home1/02064/henryiii/work_dir/files_mini_real_fto5/Det6_mini_real_fto5_mpi0_t0.root')
----> 3 d['TrackedRays']

~/.local/share/virtualenvs/build_newtacc-_nOng7Ur/lib/python3.7/site-packages/uproot/rootio.py in __getitem__(self, name)
    234 
    235     def __getitem__(self, name):
--> 236         return self.get(name)
    237 
    238     def __len__(self):

~/.local/share/virtualenvs/build_newtacc-_nOng7Ur/lib/python3.7/site-packages/uproot/rootio.py in get(self, name, cycle)
    352 
    353             if last is not None:
--> 354                 return last.get()
    355             elif cycle is None:
    356                 raise KeyError("not found: {0}".format(repr(name)))

~/.local/share/virtualenvs/build_newtacc-_nOng7Ur/lib/python3.7/site-packages/uproot/rootio.py in get(self, dismiss)
    898 
    899         try:
--> 900             return _classof(self._context, self._fClassName).read(self._source, self._cursor.copied(), self._context, self)
    901         finally:
    902             if dismiss:

~/.local/share/virtualenvs/build_newtacc-_nOng7Ur/lib/python3.7/site-packages/uproot/rootio.py in read(cls, source, cursor, context, parent)
    841         out = cls.__new__(cls)
    842         out = cls._readinto(out, source, cursor, context, parent)
--> 843         out._postprocess(source, cursor, context, parent)
    844         return out
    845 

~/.local/share/virtualenvs/build_newtacc-_nOng7Ur/lib/python3.7/site-packages/uproot/tree.py in _postprocess(self, source, cursor, context, parent)
    261 
    262         for branch in self._fBranches:
--> 263             self._attachstreamer(branch, context.streamerinfosmap.get(getattr(branch, "_fClassName", None), None), context.streamerinfosmap)
    264 
    265         self._branchlookup = {}

~/.local/share/virtualenvs/build_newtacc-_nOng7Ur/lib/python3.7/site-packages/uproot/tree.py in _attachstreamer(self, branch, streamer, streamerinfosmap)
    180 
    181             if m is None:
--> 182                 if branch.name in streamerinfosmap:
    183                     streamer = streamerinfosmap[branch.name]
    184                 else:

AttributeError: 'Undefined' object has no attribute 'name'
```

**Comment by henryiii** [comment]
I've sent you links to the files. The issue seems to be in the large branches. It is reading 265108 entries but expecting 321208 (always different by 56100, all branches fail). fUnits is set to 8 on the large file, and 4 on the small file, but doesn't seem to change much in uproot. My guess is that some of the i4 or `+ 4`'s in uproot should conditionally be i8 or `+ 8` somewhere.

**Comment by jpivarski** [comment]
PR #204.

I claim innocence. The streamer Geant produces for TBranch.fBasketSeek says that this array has 32-bit values regardless of whether the file is big or not. The fBasketSeek values are clearly 64-bit (and they have to be, to point to seek points in the file).

At first (ac7f015c63b951e4b9bb66a6e82e9ba4a07fe541), I "fixed" the streamer because I assumed that Geant's custom ROOT I/O wasn't smart enough to make the streamer match the file type. But then, that would mean that ROOT itself wouldn't be able to read the file, and it _can_.

How does ROOT do it? By [special purpose code](https://github.com/root-project/root/blob/3d7d29a949e190d2e8f5dad3da5c2c8e016b1d9e/tree/tree/src/TBranch.cxx#L2631) that overloads the `isArray` byte (0 = null pointer, 1 = non-null, but may be an empty array) with another meaning (2 = the values in this array are 64-bit). I implemented this dynamic logic (98808a0b256d0686c338db3d423cb25dd0e26b53) because in principle, two TTrees in the same file could have different seek types.

It wasn't in any `+4` or `+8` code anywhere; these are all in ROOTDirectories and TKeys that use the fVersion mechanism to know they should use 64-bit seek values. This case is in streamer-generated code that must disagree with its streamer.

**Comment by henryiii** [comment]
Wow, nice catch. I was looking through this but missed the isArray trick. Thanks!
